cpp/ql/lib/semmle/code/cpp/Element.qll

```diff
@@ -138,31 +138,40 @@ class Element extends ElementBase {
     or
     exists(TemplateClass tc | this = tc.getATemplateArgument() and result = tc)
     or
-    // result instanceof namespace
+    // result instanceof Namespace
     exists(Namespace n | result = n and n.getADeclaration() = this)
     or
     exists(FriendDecl d, Namespace n | this = d and n.getADeclaration() = d and result = n)
     or
     exists(Namespace n | this = n and result = n.getParentNamespace())
     or
-    // result instanceof stmt
+    // result instanceof Stmt
     exists(LocalVariable v |
       this = v and
       exists(DeclStmt ds | ds.getADeclaration() = v and result = ds.getParent())
     )
     or
-    exists(Parameter p | this = p and result = p.getFunction())
+    exists(Parameter p |
+      this = p and
+      (
+        result = p.getFunction() or
+        result = p.getCatchBlock().getParent().(Handler).getParent().(TryStmt).getParent() or
```

Comment:
@jketema Could you explain the rationale behind making the parent scope of a catch block parameter the parent scope of the `TryStmt`? Surely it should be the catch block itself?

This is what is breaking the Coding Standards `IdentifierHiding.ql` query. That query tries to find pairs of identifiers where the second hides the first. By considering to the parameter of a catch block to have a parent scope of the parent of the `TryStmt` we erroneously report cases like this:
```cpp
  {
    int i; // COMPLIANT
  }

  try {

  } catch (int i) { // COMPLIANT
  }
```

Reply:
> Could you explain the rationale behind making the parent scope of a catch block parameter the parent scope of the TryStmt? Surely it should be the catch block itself?

This seemed to be most consistent with the other cases here. For example, for functions the parent scope is the whole function, and not the statement block that forms the body of the function. The latter would correspond with the catch block in the case of a try-catch.

Reply:
> For example, for functions the parent scope is the whole function

Just checking, but I presume you mean "for parameters the parent scope is the whole function"?

If so, I think it would be reasonable on that basis to report the `Handler` of the `CatchBlock` as the parent scope of the parameter. But reporting the _parent_ of the `TryStmt` as the "scope" erroneously puts it in the same scope as other variables declared outside the `TryStmt`. Given the main purpose of this predicate is to support identifier hiding queries, this seems like it's not ideal.

Reply:
The problem is that `Handler` is not exposed anywhere else in the QL API, and it's only really visible in the AST, so that's far from ideal either.